Missing retry logic in curielogger's ES initialization code paths

Hi,

Testing a deployment of #168 in Helm, I encountered an error I had not obtained when testing #117: curielogger failed at `client.ILM.GetLifecycle` because ES wasn't ready yet. https://github.com/curiefense/curiefense/blob/63b64f271f313f75f1c787bd49ac444e4671f735/curiefense/curielogger/curielogger/outputs.elasticsearch.go#L383

I think adding some retry logic here could help, maybe something similar to what @flaper87 did in #168's in `ConfigureKibana`.

Same problem when running this in Helm:
```
2021/02/25 21:35:33 [ERROR] There was an error while querying the ILM Policies dial tcp 172.18.0.5:9200: connect: connection refused
```

--- a/curiefense/curielogger/curielogger.yml
+++ b/curiefense/curielogger/curielogger.yml
@@ -6,7 +6,7 @@ outputs:
         overwrite: true
         use_data_stream: true
         url: "http://elasticsearch:9200/"
-        kibana_url: "http://curiefense:5601"
+        kibana_url: "http://kibana:5601"
 
     - logstash:
         enabled: true

--- a/curiefense/curielogger/curielogger.yml
+++ b/curiefense/curielogger/curielogger.yml
@@ -6,6 +6,7 @@ outputs:
         overwrite: true
         use_data_stream: true
         url: "http://elasticsearch:9200/"
+        kibana_url: "http://curiefense:5601"
 
     - logstash:
         enabled: true

--- a/curiefense/curielogger/curielogger/main.go
+++ b/curiefense/curielogger/curielogger/main.go
@@ -569,7 +569,7 @@ func (l *logstashLogger) Configure(channel_capacity int) error {
 	if l.config.Elasticsearch.Url != "" {
 		log.Printf("[DEBUG] elasticsearch configs set, initializing configuration steps for %s", l.config.Elasticsearch.Url)
 		es := ElasticsearchLogger{config: l.config.Elasticsearch}
-		es.Configure(0)
+		return es.Configure(0)
 	}
 
 	return nil
@@ -928,21 +928,33 @@ func main() {
 		loggers = append(loggers, &pg)
 	}
 
+	configRetry := func(logger Logger) {
+		for i := 0; i < 60; i++ {
+			err := logger.Configure(config.ChannelCapacity)
+
+			if err == nil {
+				loggers = append(loggers, logger)
+				break
+			}
+
+			log.Printf("[ERROR]: failed to configure logger (retrying in 5s) %v %v", logger, err)
+			time.Sleep(5 * time.Second)
+		}
+	}
+
 	for _, output := range config.Outputs {
 		// ElasticSearch
 		if output.Elasticsearch.Enabled {
 			log.Printf("[DEBUG] Elasticsearch enabled with URL: %s", output.Elasticsearch.Url)
 			es := ElasticsearchLogger{config: output.Elasticsearch}
-			es.Configure(config.ChannelCapacity)
-			loggers = append(loggers, &es)
+			go configRetry(&es)
 		}
 
 		// Logstash
 		if output.Logstash.Enabled {
 			log.Printf("[DEBUG] Logstash enabled with URL: %s", output.Logstash.Url)
 			ls := logstashLogger{config: output.Logstash}
-			ls.Configure(config.ChannelCapacity)
-			loggers = append(loggers, &ls)
+			go configRetry(&ls)
 		}
 
 	}

--- a/curiefense/curielogger/curielogger/outputs.elasticsearch.go
+++ b/curiefense/curielogger/curielogger/outputs.elasticsearch.go
@@ -6,6 +6,7 @@ import (
 	"fmt"
 	"html/template"
 	"log"
+	"net/http"
 	"strings"
 	"time"
 
@@ -326,6 +327,15 @@ const (
 }
 	`
 
+	KIBANA_INDEX_PATTERN = `
+{
+  "attributes": {
+    "timeFieldName": "@timestamp",
+    "title": "{{.AccessLogIndexName}}*"
+  }
+}
+	`
+
 	METRICS_ES_PREFIX   = "curiemetrics"
 	ACCESSLOG_ES_PREFIX = "curieaccesslog"
 )
@@ -333,6 +343,7 @@ const (
 type ElasticsearchConfig struct {
 	Enabled            bool   `mapstructure:"enabled"`
 	Url                string `mapstructure:"url"`
+	KibanaUrl          string `mapstructure:"kibana_url"`
 	Initialize         bool   `mapstructure:"initialize"`
 	Overwrite          bool   `mapstructure:"overwrite"`
 	AccessLogIndexName string `mapstructure:"accesslog_index_name"`
@@ -360,6 +371,49 @@ func (l ElasticsearchLogger) getESClient() *elasticsearch.Client {
 		time.Sleep(time.Second)
 	}
 	return l.client
+}
+
+func (l *ElasticsearchLogger) ConfigureKibana() {
+
+	var ktpl bytes.Buffer
+	gTpl := template.Must(template.New("it").Parse(KIBANA_INDEX_PATTERN))
+	gTpl.Execute(&ktpl, l.config)
+
+	body := bytes.NewReader(ktpl.Bytes())
+
+	log.Printf("[DEBUG]: configuring kibana")
+	kbUrl := fmt.Sprintf("%s/api/saved_objects/index-pattern/%s*", l.config.KibanaUrl, l.config.AccessLogIndexName)
+
+	client := http.Client{
+		Timeout: time.Duration(5 * time.Second),
+	}
+
+	req, err := http.NewRequest("POST", kbUrl, body)
+
+	if err != nil {
+		return
+	}
+
+	req.Header.Set("Content-Type", "application/json")
+	req.Header.Set("kbn-xsrf", "true")
+
+	for i := 0; i < 60; i++ {
+
+		rst, err := client.Do(req)
+
+		if err == nil && rst.StatusCode == 200 {
+			log.Printf("[INFO]: kibana index pattern created %s", kbUrl)
+			break
+		}
+
+		if rst.StatusCode == 409 {
+			log.Printf("[INFO]: kibana index pattern already exists %s", kbUrl)
+			break
+		}
+
+		log.Printf("[ERROR]: kibana index pattern creation failed (retrying in 5s) %s %v %v", kbUrl, err, rst)
+		time.Sleep(5 * time.Second)
+	}
 }
 
 func (l *ElasticsearchLogger) Configure(channel_capacity int) error {
@@ -465,6 +519,10 @@ func (l *ElasticsearchLogger) Configure(channel_capacity int) error {
 			log.Printf("[ERROR] index template creation failed %v %v", err, resp)
 		}
 	}
+
+	// Attempt to configure Kibana's index patterns
+	// and dashboards in the background.
+	go l.ConfigureKibana()
 
 	return nil
 }

--- a/curiefense/curielogger/curielogger/outputs.elasticsearch.go
+++ b/curiefense/curielogger/curielogger/outputs.elasticsearch.go
@@ -401,14 +401,16 @@ func (l *ElasticsearchLogger) ConfigureKibana() {
 
 		rst, err := client.Do(req)
 
-		if err == nil && rst.StatusCode == 200 {
-			log.Printf("[INFO]: kibana index pattern created %s", kbUrl)
-			break
-		}
+		if rst != nil {
+			if rst.StatusCode == 200 {
+				log.Printf("[INFO]: kibana index pattern created %s", kbUrl)
+				break
+			}
 
-		if rst.StatusCode == 409 {
-			log.Printf("[INFO]: kibana index pattern already exists %s", kbUrl)
-			break
+			if rst.StatusCode == 409 {
+				log.Printf("[INFO]: kibana index pattern already exists %s", kbUrl)
+				break
+			}
 		}
 
 		log.Printf("[ERROR]: kibana index pattern creation failed (retrying in 5s) %s %v %v", kbUrl, err, rst)

--- a/curiefense/curielogger/curielogger/outputs.elasticsearch.go
+++ b/curiefense/curielogger/curielogger/outputs.elasticsearch.go
@@ -440,7 +440,7 @@ func (l *ElasticsearchLogger) Configure(channel_capacity int) error {
 
 	if err != nil {
 		log.Printf("[ERROR] There was an error while querying the ILM Policies %v", err)
-		return nil
+		return err
 	}
 
 	var ilm map[string]json.RawMessage
@@ -475,7 +475,7 @@ func (l *ElasticsearchLogger) Configure(channel_capacity int) error {
 
 	if err != nil {
 		log.Printf("[ERROR] there was an error while querying the template %v", err)
-		return nil
+		return err
 	}
 
 	if l.config.Overwrite || tplExists.IsError() {
@@ -504,7 +504,7 @@ func (l *ElasticsearchLogger) Configure(channel_capacity int) error {
 
 		if err != nil {
 			log.Printf("[ERROR] there was an error while querying the template %v", err)
-			return nil
+			return err
 		}
 
 		if !iExists.IsError() {
@@ -519,6 +519,7 @@ func (l *ElasticsearchLogger) Configure(channel_capacity int) error {
 		resp, err := client.Indices.Create(indexName, client.Indices.Create.WithBody(bytes.NewReader(iTpl.Bytes())))
 		if err != nil || resp.IsError() {
 			log.Printf("[ERROR] index template creation failed %v %v", err, resp)
+			return err
 		}
 	}
 

--- a/curiefense/images/curielogger/init/start_curielogger.sh
+++ b/curiefense/images/curielogger/init/start_curielogger.sh
@@ -1,8 +1,5 @@
 #!/bin/bash
 
-ES_INDEX_NAME=${CURIEFENSE_ES_INDEX_NAME:-curieaccesslog}
-ES_PATTERN_PATH="api/saved_objects/index-pattern/${ES_INDEX_NAME}"
-
 CURL="curl --silent --write-out %{http_code}\n -H Content-Type:application/json --output /dev/null"
 
 wait_for_es () {
@@ -12,39 +9,9 @@ wait_for_es () {
 	fi
 }
 
-create_kibana_index_pattern () {
-	# Wait for kibana to become reachable
-	while true; do
-		if $CURL "${CURIELOGGER_KIBANA_URL}api/status"|grep -q 200; then
-			break
-		fi
-		echo "Kibana at $CURIELOGGER_KIBANA_URL is not reachable yet, waiting 5s..."
-		sleep 5
-	done
-
-	# Check whether the index pattern already exists
-	if $CURL "$CURIELOGGER_KIBANA_URL$ES_PATTERN_PATH"|grep -q 200; then
-		# already exists
-		echo "Kibana index already exists."
-	else
-		# Create the index pattern
-		if $CURL -X POST "$CURIELOGGER_KIBANA_URL$ES_PATTERN_PATH" -H 'kbn-xsrf: true' -d '{"attributes": {"title": "'"$ES_INDEX_NAME*"'","timeFieldName": "@timestamp"}}'|grep -q 200; then
-
-			echo "Kibana index $ES_INDEX_NAME created."
-		else
-			echo "Kibana index $ES_INDEX_NAME creation failed, retrying."
-			sleep 5
-			create_kibana_index_pattern
-		fi
-	fi
-}
-
-
-if [ -n "$CURIELOGGER_INITIALIZE_KIBANA_INDEX" ]; then
-	# in case logs are saved in elasticsearch and not postgres
-	>&2 echo "Creating an index pattern in Kibana if needed."
+if [ -n "$CURIELOGGER_WAIT_FOR_ES" ]; then
+	>&2 echo "Waiting for Elasticsearch"
 	wait_for_es
-	create_kibana_index_pattern
 fi
 
 /bin/curielogger

